specification/window.browser.bs

```diff
@@ -0,0 +1,80 @@
+<pre class="metadata">
+Title: window.browser
+Shortname: wecg-browser
+Level: 1
+Group: wecg
+Status: UD
+URL: https://w3c.github.io/webextensions/specification/window.browser.html
+Editor: Patrick Kettner, Google, patrickkettner@google.com
+Abstract: This specification reserves the <code>window.browser</code> namespace for use by WebExtensions.
+Repository: w3c/webextensions
+Complain About: accidental-2119 on, missing-example-ids on
+</pre>
+
+<pre class=link-defaults>
+spec:html; type:attribute; text:window
+</pre>
+
+<pre class="anchors">
+spec: html; urlPrefix: https://html.spec.whatwg.org/multipage/
+    type: dfn; text: valid non-empty URL; url: valid-non-empty-url-potentially-surrounded-by-spaces
+</pre>
+
+<section class="non-normative">
+  ## Introduction {#introduction}
+  <em>This section is non-normative.</em>
+
+  Browser extensions are programs to extend the capabilities of web browsers.
+  Although not standardized at first, web browsers have independently implemented a common interoperable set of extension APIs, called WebExtensions.
+
+  Modern WebExtensions are an evolution of a system originally introduced in
+  Chrome version 3. In that system, any API that was added was put under the
+  <code>chrome</code> namespace. When WebExtensions were adopted outside of
+  Chromium, other vendors used the more neutral <code>browser</code> namespace.
+  That <code>browser</code> is now considered to canonically correct namespace
+  for all WebExtension APIs.  Some interactions with WebExtensions (such as
+  external messaging from content scripts to background processes) necessitate
+  the exposure of this namespace to the web.
+
+  Given that, this document defines <code>window.browser</code> as reserved for
+  the use of APIs designed to interact with WebExtensions. The specifics of what
+  APIs within <code>window.browser</code> are at present intentionally undefined,
+  and are left up to User Agent implementation.
+</section>
+
+
+  <section>
+    <h3 id="window-interface">
+      <a attribute lt="browser"><code>window.browser</code></a> API
+    </h3>
+
+    {{browser}} is UA defined attribute exposed on {{window}}. When implemented,
+    it MUST be used only for WebExtension related functionality.
+  </section>
+
+
+<pre class="idl">
+  partial interface Window {
+   attribute object browser;
```

Comment:
Should we make it clear that this is sometimes undefined, such as if a site is not included in the `externally_connectable` array in any extension's manifest? I'm guessing the right syntax based on [this](https://hg.mozilla.org/integration/autoland/file/d65760500d063d80f6cc3c1fa8c61bfe1ac81412/dom/bindings/test/TestExampleGen.webidl#l582) but it seems to be a new feature that doesn't appear in many real specs so not sure if there's a better way. cc/ @Rob--W @xeenon 

```suggestion
   attribute (undefined or object) browser;
```

Reply:
it feels weird to set the only attribute being specified to undefined in the spec. Certainly can add it, but I am not sure I see the value if it is user agent defined. Would love to hear Rob's and Timothy's thoughts

Reply:
Since WebIDL treats `object` and `undefined` as distinct types, it seems like this would more accurately represent the possible values a user agent may assign the global we're reserving. 

Would it be more appropriate (or conventional) to use a union type as @oliverdunk suggested, a nullable type, or possibly even to declare a (partial) namespace?

```
// Nullable object
partial interface Window {
 attribute object? browser;
};
```

```
// Partial namespace
[Exposed=(Window,Worker)]
partial namespace browser {}
```

I tried looking around at other specs to see if/when/how `namespace` was used and was surprised at how rarely it appeared. The only notable place I saw it used was the [console spec](https://console.spec.whatwg.org/#console-namespace).

Reply:
As a note, my (uneducated) reading of the spec was that nullable types were equivalent to `null | something` so I'm not sure they'd be appropriate here. Still very unsure what the right path is though.

Reply:
We've agreed to revisit this after landing.

Reply:
@oliverdunk can you LGTM?